fix: Enable default triggers for argo rollouts

RaviHari opened the pull request:
Fixes: #1668 

Signed-off-by: Ravi Hari <ravireliable@gmail.com>

Checklist:

* [ X ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ X ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ X ] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ X ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ X ] My builds are green. Try syncing with master if they are not. 
* [ X ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

huikang commented on utils/record/record.go:
```diff
@@ -1,13 +1,18 @@
-	"github.com/argoproj/notifications-engine/pkg/services"
-
-	"github.com/argoproj/notifications-engine/pkg/api"
+	rolloutapi "github.com/argoproj/notifications-engine/pkg/api"
```
should rolloutapi be `notificationapi`?

huikang commented on utils/record/record.go:
```diff
@@ -1,13 +1,18 @@
+	"github.com/argoproj/notifications-engine/pkg/services"
+	"k8s.io/client-go/tools/cache"
+
+	//"github.com/argoproj/notifications-engine/pkg/triggers"
```
remove line 9?

RaviHari commented:
Testing Done:

Logs:
```
INFO[2021-12-08T22:33:14+05:30] Start processing                              resource=rollout-demo/rollout-bluegreen
INFO[2021-12-08T22:33:14+05:30] Trigger on-rollout-completed result: [{[0].xxxxx  [rollout-completed] false}]  resource=rollout-demo/rollout-bluegreen
INFO[2021-12-08T22:33:14+05:30] Trigger on-scaling-replica-set result: [{[0].xxxxx  [scaling-replicaset] false}]  resource=rollout-demo/rollout-bluegreen
```

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: argo-rollouts-notification-configmap
  namespace: argo-rollouts
data:
  service.slack: |
    token: $slack-token
    icon: https://avatars.githubusercontent.com/u/30269780?s=400&v=4
  template.rollout-completed: |
    message: Rollout {{.rollout.metadata.name}} has been completed.
    slack:
      attachments: |
          [{
            "title": "{{ .rollout.metadata.name}}",
            "color": "#18be52",
            "fields": [
            {
              "title": "Strategy",
              "value": "{{if .rollout.spec.strategy.blueGreen}}BlueGreen{{end}}{{if .rollout.spec.strategy.canary}}Canary{{end}}",
              "short": true
            }
            {{range $index, $c := .rollout.spec.template.spec.containers}}
              {{if not $index}},{{end}}
              {{if $index}},{{end}}
              {
                "title": "{{$c.name}}",
                "value": "{{$c.image}}",
                "short": true
              }
            {{end}}
            ]
          }]
  template.rollout-step-completed: |
    message: Rollout {{.rollout.metadata.name}} step number {{ add .rollout.status.currentStepIndex 1}}/{{len .rollout.spec.strategy.canary.steps}} has been completed.
    email:
      subject: Rollout {{.rollout.metadata.name}} step number {{ add .rollout.status.currentStepIndex 1}}/{{len .rollout.spec.strategy.canary.steps}} has been completed.
    slack:
      attachments: |
          [{
            "title": "{{ .rollout.metadata.name}}",
            "color": "#18be52",
            "fields": [
            {
              "title": "Strategy",
              "value": "{{if .rollout.spec.strategy.blueGreen}}BlueGreen{{end}}{{if .rollout.spec.strategy.canary}}Canary{{end}}",
              "short": true
            },
            {
              "title": "Step completed",
              "value": "{{add .rollout.status.currentStepIndex 1}}/{{len .rollout.spec.strategy.canary.steps}}",
              "short": true
            }
            {{range $index, $c := .rollout.spec.template.spec.containers}}
              {{if not $index}},{{end}}
              {{if $index}},{{end}}
              {
                "title": "{{$c.name}}",
                "value": "{{$c.image}}",
                "short": true
              }
            {{end}}
            ]
          }]
  template.rollout-updated: |
    message: Rollout {{.rollout.metadata.name}} has been updated.
    email:
      subject: Rollout {{.rollout.metadata.name}} has been updated.
    slack:
      attachments: |
          [{
            "title": "{{ .rollout.metadata.name}}",
            "color": "#18be52",
            "fields": [
            {
              "title": "Strategy",
              "value": "{{if .rollout.spec.strategy.blueGreen}}BlueGreen{{end}}{{if .rollout.spec.strategy.canary}}Canary{{end}}",
              "short": true
            }
            {{range $index, $c := .rollout.spec.template.spec.containers}}
              {{if not $index}},{{end}}
              {{if $index}},{{end}}
              {
                "title": "{{$c.name}}",
                "value": "{{$c.image}}",
                "short": true
              }
            {{end}}
            ]
          }]
  template.scaling-replicaset: |
    message: Scaling Rollout {{.rollout.metadata.name}}'s replicaset to {{.rollout.spec.replicas}}.
    email:
      subject: Scaling Rollout {{.rollout.metadata.name}}'s replcaset to {{.rollout.spec.replicas}}.
    slack:
      attachments: |
          [{
            "title": "{{ .rollout.metadata.name}}",
            "color": "#18be52",
            "fields": [
            {
              "title": "Strategy",
              "value": "{{if .rollout.spec.strategy.blueGreen}}BlueGreen{{end}}{{if .rollout.spec.strategy.canary}}Canary{{end}}",
              "short": true
            },
            {
              "title": "Desired replica",
              "value": "{{.rollout.spec.replicas}}",
              "short": true
            },
            {
              "title": "Updated replicas",
              "value": "{{.rollout.status.updatedReplicas}}",
              "short": true
            }
            {{range $index, $c := .rollout.spec.template.spec.containers}}
              {{if not $index}},{{end}}
              {{if $index}},{{end}}
              {
                "title": "{{$c.name}}",
                "value": "{{$c.image}}",
                "short": true
              }
            {{end}}
            ]
          }]
  trigger.on-rollout-completed: |
    - send: [rollout-completed]
  trigger.on-rollout-step-completed: |
    - send: [rollout-step-completed]
  trigger.on-rollout-updated: |
    - send: [rollout-updated]
  trigger.on-scaling-replica-set: |
    - send: [scaling-replicaset]
  defaultTriggers: |
    - on-scaling-replica-set
```

slack messages:
```

testAPP  10:40 PM
Scaling Rollout rollout-bluegreen's replicaset to 1.
rollout-bluegreen
Strategy
BlueGreen
Desired replica
1
Updated replicas
2
rollouts-demo
argoproj/rollouts-demo:blue


New
10:40
Scaling Rollout rollout-bluegreen's replicaset to 2.
rollout-bluegreen
Strategy
BlueGreen
Desired replica
2
Updated replicas
1
rollouts-demo
argoproj/rollouts-demo:blue


10:41
Scaling Rollout rollout-bluegreen's replicaset to 3.
rollout-bluegreen
Strategy
BlueGreen
Desired replica
3
Updated replicas
2
rollouts-demo
argoproj/rollouts-demo:blue
```

huikang commented on utils/record/record.go:
```diff
@@ -89,6 +94,39 @@ type FakeEventRecorder struct {
 
+func NewFakeApiFactory() api.Factory {
+	var (
+		settings = api.Settings{ConfigMapName: "my-config-map", SecretName: "my-secret", InitGetVars: func(cfg *api.Config, configMap *corev1.ConfigMap, secret *corev1.Secret) (api.GetVars, error) {
```
where are the names `my-config-map` and `my-secret` coming from?

Same question on line 107, 113.

RaviHari replied:
Ideally they should be coming from controller.go
but as we are creating here FakeApiFactory for testing giving in the values here directly.. similar to that of notification engine.

huikang commented on utils/record/record.go:
```diff
@@ -167,23 +204,21 @@ func NewAPIFactorySettings() api.Settings {
-	}
-
 	// Creates config for notifications for built-in triggers
 	triggerActions, ok := notificationsAPI.GetConfig().Triggers[trigger]
```
notificationsAPI.GetConfig() can be replaced by the variable cfg as defined in line 211

huikang commented on utils/record/record.go:
```diff
@@ -98,7 +135,7 @@ func NewFakeEventRecorder() *FakeEventRecorder {
 			[]string{"name", "namespace", "type", "reason"},
 		),
-		nil,
+		NewFakeApiFactory(),
```
Is there any reason to introduce this `NewFakeApiFactory()`? Thanks.

RaviHari replied:
If we donot use this we will get nil access reference error in line 207

jessesuen commented:
@khhirani do you mind taking a look at this?

RaviHari commented:
I noticed that some times this test-e2e fails on test case: TestIstioSuite/TestIstioAbortUpdateDeleteAllCanaryPods as given here:  https://github.com/argoproj/argo-rollouts/runs/4494296112?check_suite_focus=true I think this is primarily due to a timeout not a functional defect. 

When I run this test locally it works fine:

```
 make test-e2e E2E_TEST_OPTIONS="-testify.m TestIstioAbortUpdateDeleteAllCanaryPods"   
go test -timeout 30m -v -count 1 --tags e2e -p 4 --short ./test/e2e -testify.m TestIstioAbortUpdateDeleteAllCanaryPods
=== RUN   TestAnalysisSuite
    suite.go:190: warning: no tests to run
--- PASS: TestAnalysisSuite (0.00s)
=== RUN   TestAWSSuite
    suite.go:190: warning: no tests to run
--- PASS: TestAWSSuite (0.00s)
=== RUN   TestBlueGreenSuite
    suite.go:190: warning: no tests to run
--- PASS: TestBlueGreenSuite (0.00s)
=== RUN   TestCanarySuite
    suite.go:190: warning: no tests to run
--- PASS: TestCanarySuite (0.00s)
=== RUN   TestExperimentSuite
    suite.go:190: warning: no tests to run
--- PASS: TestExperimentSuite (0.00s)
=== RUN   TestFunctionalSuite
    suite.go:190: warning: no tests to run
--- PASS: TestFunctionalSuite (0.00s)
=== RUN   TestIstioSuite
=== RUN   TestIstioSuite/TestIstioAbortUpdateDeleteAllCanaryPods
time="2021-12-13T13:13:40+05:30" level=info msg="Deleting e2e-test-name=TestIstioAbortUpdateDeleteAllCanaryPods"
time="2021-12-13T13:13:43+05:30" level=info msg="service/istio-host-split-canary created\n" rollout=istio-host-split
time="2021-12-13T13:13:43+05:30" level=info msg="service/istio-host-split-stable created\n" rollout=istio-host-split
time="2021-12-13T13:13:43+05:30" level=info msg="virtualservice.networking.istio.io/istio-host-split-vsvc created\n" rollout=istio-host-split
time="2021-12-13T13:13:43+05:30" level=info msg="rollout.argoproj.io/istio-host-split created\n" rollout=istio-host-split
time="2021-12-13T13:13:43+05:30" level=info msg="Waiting for condition: status=Healthy" rollout=istio-host-split
time="2021-12-13T13:13:54+05:30" level=info msg="Condition 'status=Healthy' met after 10s" rollout=istio-host-split
time="2021-12-13T13:13:54+05:30" level=info msg="Updated rollout pod spec: 2021-12-13T13:13:54.128467+05:30" rollout=istio-host-split
time="2021-12-13T13:13:54+05:30" level=info msg="Waiting for condition: status=Paused" rollout=istio-host-split
time="2021-12-13T13:13:59+05:30" level=info msg="Condition 'status=Paused' met after 5s" rollout=istio-host-split
time="2021-12-13T13:13:59+05:30" level=info msg="Pod expectation 'revision:2 pod count == 2' met" rollout=istio-host-split
time="2021-12-13T13:13:59+05:30" level=info msg="Promoted rollout" rollout=istio-host-split
time="2021-12-13T13:13:59+05:30" level=info msg="Waiting for condition: status=Paused" rollout=istio-host-split
time="2021-12-13T13:13:59+05:30" level=info msg="Condition 'status=Paused' met after 0s" rollout=istio-host-split
time="2021-12-13T13:13:59+05:30" level=info msg="Promoted rollout" rollout=istio-host-split
time="2021-12-13T13:13:59+05:30" level=info msg="Waiting for condition: status=Paused" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Condition 'status=Paused' met after 4s" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Pod expectation 'revision:2 pod count == 4' met" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Aborted rollout" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Waiting for condition: status=Degraded" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Condition 'status=Degraded' met after 0s" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Pod expectation 'revision:1 pod count == 5' met" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Pod expectation 'revision:2 pod count == 4' met" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="rs expectation 'RS istio-host-split-b8d98f5fd scale down' met" rollout=istio-host-split
time="2021-12-13T13:14:04+05:30" level=info msg="Deleting e2e-test-name=TestIstioAbortUpdateDeleteAllCanaryPods"
time="2021-12-13T13:14:04+05:30" level=info msg="Deleting e2e-test-name"
--- PASS: TestIstioSuite (24.76s)
    --- PASS: TestIstioSuite/TestIstioAbortUpdateDeleteAllCanaryPods (24.49s)
=== RUN   TestSMIIngressSuite
    suite.go:190: warning: no tests to run
--- PASS: TestSMIIngressSuite (0.00s)
=== RUN   TestSMISuite
    suite.go:190: warning: no tests to run
--- PASS: TestSMISuite (0.00s)
PASS
ok      github.com/argoproj/argo-rollouts/test/e2e      25.688s

```

jessesuen commented:
> I think this is primarily due to a timeout not a functional defect.

Yes tests are flaky right now. I'll resubmit it when I think/notice it is the case.

harikrongali commented:
@RaviHari let me validate and come back. quick question, have you applied https://github.com/argoproj/argo-rollouts/blob/master/manifests/notifications-install.yaml ?

RaviHari commented:
> @RaviHari let me validate and come back. quick question, have you applied https://github.com/argoproj/argo-rollouts/blob/master/manifests/notifications-install.yaml ?

@harikrongali ,
Thanks for loking into it.. In my local test yes I applied.. but the E2E tests now fail for a different usecase.. `FAIL: TestFunctionalSuite/TestBlueGreenUpdate` which passed  in earlier runs.

jessesuen reviewed (approved):
LGTM!

alexmt commented on utils/record/record.go:
```diff
@@ -167,25 +206,23 @@ func NewAPIFactorySettings() api.Settings {
 func (e *EventRecorderAdapter) sendNotifications(object runtime.Object, opts EventOptions) error {
+	notificationsAPI, err := e.apiFactory.GetAPI()
+	if err != nil {
+		return err
```
One minor issue here. We are going to pollute controller logs with error messages if notifications are not configured at all.  The error message should be printed only if someone explicitly added subscribe annotation and notifications are not configured. You can use the following snippet to avoid printing unnecessary error message:

```golang
import "k8s.io/apimachinery/pkg/api/errors"
```
```golang
// don't return error if notifications are not configured and rollout has no subscribers
subsFromAnnotations := subscriptions.Annotations(object.(metav1.Object).GetAnnotations())
if errors.IsNotFound(err) && len(subsFromAnnotations.GetDestinations(nil, map[string][]string{})) == 0 {
	return nil
}
return err
```

RaviHari replied:
sure sounds good.

alexmt reviewed (approved):
LGTM
